packages/pwa-buildpack/lib/WebpackTools/plugins/ServiceWorkerPlugin.js

```diff
@@ -1,34 +1,54 @@
-// TODO: (p1) write test file and test
+const path = require('path');
+const webpack = require('webpack');
 const WorkboxPlugin = require('workbox-webpack-plugin');
 const WriteFileWebpackPlugin = require('write-file-webpack-plugin');
+
+const createFileHash = require('../../Utilities/createFileHash');
 const optionsValidator = require('../../util/options-validator');
 
 // No longer modifiable since v4
-const SW_FILENAME = 'sw.js';
+let SERVICE_WORKER_FILENAME = 'sw.js';
+let SERVICE_WORKER_HASH = '';
+
+const DEFAULT_SW_FILE_PATH = path.join(
+    path.resolve(__dirname).split('/packages')[0],
+    'packages',
+    'venia-concept',
+    'src',
+    'sw.js'
+);
+
+createFileHash(DEFAULT_SW_FILE_PATH).then(hash => {
+    SERVICE_WORKER_HASH = hash;
+    SERVICE_WORKER_FILENAME = `sw.${hash}.js`;
+});
 
 class ServiceWorkerPlugin {
     constructor(config) {
         ServiceWorkerPlugin.validateOptions('ServiceWorkerPlugin', config);
         this.config = config;
     }
+
     applyGenerateSW(compiler) {
         const config = {
             // `globDirectory` and `globPatterns` must match at least 1 file
             // otherwise workbox throws an error
             globDirectory: this.config.paths.output,
+
             // TODO: (feature) autogenerate glob patterns from asset manifest
             globPatterns: ['**/*.{gif,jpg,png,svg}'],
 
             // activate the worker as soon as it reaches the waiting phase
             skipWaiting: true,
 
             // the max scope of a worker is its location
-            swDest: SW_FILENAME
+            swDest: SERVICE_WORKER_FILENAME
```

Comment:
What does the comment above this line mean? What is "the max scope"?

Reply:
That was written by Zetlen a while ago but here is my take on that. According to the SW docs:

[One subtlety with the register() method is the location of the service worker file. You'll notice in this case that the service worker file is at the root of the domain. This means that the service worker's scope will be the entire origin. In other words, this service worker will receive fetch events for everything on this domain. If we register the service worker file at /example/sw.js, then the service worker would only see fetch events for pages whose URL starts with /example/ (i.e. /example/page1/, /example/page2/).](https://developers.google.com/web/fundamentals/primers/service-workers#register_a_service_worker)

So if we decode the above paragraph and apply to our code, we are setting the domain of the service worker as the location of the default service worker file which in our case is `venia-concept/src/sw.js`.